### ISSUE TITLE
RawKeyboard triggers incorrect KeyUp when two keys are pressed and auto-repeating on web

### ISSUE THREAD

**Issue by erickzanardo** [flutter/flutter#58727]
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update your main with the following:

```
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';


void main() {
  runApp(
    MaterialApp(
      home: MyWidget(),
    ),
  );

   RawKeyboard.instance.addListener((event) {
    final isKeyDown = event is RawKeyDownEvent;
    final key = event.logicalKey.keyLabel;

    print('$isKeyDown - $key - ${event.runtimeType}');
  });
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Text(
          'Build some widgets!',
          style: Theme.of(context).textTheme.headline4,
        ),
      ),
    );
  }
}
```

3. `flutter run -d Chrome`
4. When Chrome opens, press, and keep pressed W (or any other key), then press other key. After a while release both keys.

**Expected results:**

On the console:

```
true - w - RawKeyDownEvent
true - d - RawKeyDownEvent
false - w - RawKeyUpEvent
false - d - RawKeyUpEvent
```

**Actual results:**

TLDR: Key keeps triggering the event when pressed, and when a another is pressed, the previous key gets a RawKeyUpEvent, even if it is still pressed.

```
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
false - w - RawKeyUpEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
false - d - RawKeyUpEvent
false - w - RawKeyUpEvent
```

verbose log is on attachment
[verbose.log](https://github.com/flutter/flutter/files/4733022/verbose.log)

```
flutter analyze
Analyzing bug...
No issues found! (ran in 6.1s)
```

```
flutter doctor -v
[✓] Flutter (Channel master, 1.19.0-2.0.pre, on Mac OS X 10.15.3 19D76, locale pt-BR)
    • Flutter version 1.19.0-2.0.pre at /Users/erick/flutter
    • Framework revision 9d58a87066 (13 days ago), 2020-05-22 22:37:01 -0700
    • Engine revision 9ce1e5c5c7
    • Dart version 2.9.0 (build 2.9.0-10.0.dev 7706afbcf5)


[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/erick/Library/Android/sdk
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /Users/erick/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[!] VS Code (version 1.44.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (3 available)
    • macOS      • macOS      • darwin-x64     • Mac OS X 10.15.3 19D76
    • Web Server • web-server • web-javascript • Flutter Tools
    • Chrome     • chrome     • web-javascript • Google Chrome 83.0.4103.61

! Doctor found issues in 2 categories.
```

**Comment by iapicca** [comment]
Hi @erickzanardo 
I tested your code with both the latest beta and the latest master;
I cannot reproduce the issue you are describing

<details>
  <summary>BETA logs</summary>

```bash
[+13353 ms] true - w - RawKeyDownEvent

[ +470 ms] true - w - RawKeyDownEvent

[  +86 ms] true - w - RawKeyDownEvent

[  +82 ms] true - w - RawKeyDownEvent

[  +84 ms] true - w - RawKeyDownEvent

[  +86 ms] true - w - RawKeyDownEvent

[  +81 ms] true - w - RawKeyDownEvent

[  +85 ms] true - w - RawKeyDownEvent

[  +82 ms] true - w - RawKeyDownEvent

[  +87 ms] true - w - RawKeyDownEvent

[  +81 ms] true - w - RawKeyDownEvent

[  +83 ms] true - w - RawKeyDownEvent

[  +84 ms] true - w - RawKeyDownEvent

[  +85 ms] true - w - RawKeyDownEvent

[  +81 ms] true - w - RawKeyDownEvent

[  +84 ms] true - w - RawKeyDownEvent

[  +84 ms] true - w - RawKeyDownEvent

[  +25 ms] true - p - RawKeyDownEvent

[ +114 ms] false - p - RawKeyUpEvent

[ +276 ms] false - w - RawKeyUpEvent
```

</details>


<details>
  <summary>BETA doctor</summary>

```console
[✓] Flutter (Channel beta, 1.18.0-11.1.pre, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.18.0-11.1.pre at /Users/nevercode/development/flutter_beta
    • Framework revision 2738a1148b (3 weeks ago), 2020-05-13 15:24:36 -0700
    • Engine revision ef9215ceb2
    • Dart version 2.9.0 (build 2.9.0-8.2.beta)


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.9.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (2 available)
    • Web Server • web-server • web-javascript • Flutter Tools
    • Chrome     • chrome     • web-javascript • Google Chrome 83.0.4103.61

• No issues found!
```

</details>

<details>
  <summary>MASTER logs</summary>

```bash
[ +187 ms] Launching lib/main.dart on Chrome in debug mode...
[  +66 ms] Updating assets
[ +195 ms] Syncing files to device Chrome...
[  +70 ms] Generating /Users/nevercode/Desktop/projects/master/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
[  +39 ms] <- reset
[  +21 ms] /Users/nevercode/development/flutter_master/bin/cache/dart-sdk/bin/dart
/Users/nevercode/development/flutter_master/bin/cache/artifacts/engine/darwin-x64/frontend_server.dart.snapshot --sdk-root
/Users/nevercode/development/flutter_master/bin/cache/flutter_web_sdk/ --incremental --target=dartdevc --debugger-module-names -Ddart.developer.causal_async_stacks=true
--output-dill /var/folders/c3/l5jpznsj5k37yrj47l__xpw40000gn/T/flutter_tools.MsOUFX/flutter_tool.AFAOvV/app.dill --libraries-spec
file:///Users/nevercode/development/flutter_master/bin/cache/flutter_web_sdk/libraries.json --packages .packages -Ddart.vm.profile=false -Ddart.vm.product=false
--bytecode-options=source-positions,local-var-info,debugger-stops,instance-field-initializers,keep-unreachable-code,avoid-closure-call-instructions --enable-asserts
--track-widget-creation --filesystem-root /var/folders/c3/l5jpznsj5k37yrj47l__xpw40000gn/T/flutter_tools.MsOUFX/flutter_tools.yFVsRZ --filesystem-scheme org-dartlang-app
--initialize-from-dill build/cache.dill.track.dill --platform file:///Users/nevercode/development/flutter_master/bin/cache/flutter_web_sdk/kernel/flutter_ddc_sdk.dill
[  +12 ms] <- compile org-dartlang-app:/web_entrypoint.dart
[+22236 ms] Syncing files to device Chrome... (completed in 22,378ms, longer than expected)
[   +2 ms] Synced 22.7MB.
[   +1 ms] <- accept
[   +1 ms] Caching compiled dill
[ +927 ms] [CHROME]:
[   +2 ms] [CHROME]:DevTools listening on ws://127.0.0.1:52479/devtools/browser/0becd22d-51bb-4e94-8a4d-a79dea2ea7b7
[+5078 ms] Debug service listening on ws://127.0.0.1:52512/bPOwrqaXVqc=

[ +618 ms] Debug service listening on ws://127.0.0.1:52512/bPOwrqaXVqc=
[   +2 ms] Warning: Flutter's support for web development is not stable yet and hasn't
[        ] been thoroughly tested in production environments.
[        ] For more information see https://flutter.dev/web
[        ] 🔥  To hot restart changes while running, press "r" or "R".
[   +1 ms] For a more detailed help message, press "h". To quit, press "q".
[+3021 ms] true - w - RawKeyDownEvent

[ +423 ms] true - w - RawKeyDownEvent

[  +84 ms] true - w - RawKeyDownEvent

[  +83 ms] true - w - RawKeyDownEvent

[  +83 ms] true - w - RawKeyDownEvent

[  +84 ms] true - w - RawKeyDownEvent

[  +83 ms] true - w - RawKeyDownEvent

[  +82 ms] true - w - RawKeyDownEvent

[  +87 ms] true - w - RawKeyDownEvent

[  +79 ms] true - w - RawKeyDownEvent

[  +84 ms] true - w - RawKeyDownEvent

[  +87 ms] true - w - RawKeyDownEvent

[  +82 ms] true - w - RawKeyDownEvent

[  +80 ms] true - w - RawKeyDownEvent

[  +41 ms] true - p - RawKeyDownEvent

[ +139 ms] false - p - RawKeyUpEvent

[ +539 ms] false - w - RawKeyUpEvent
```

</details>


<details>
  <summary>MASTER doctor</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 1.19.0-4.0.pre.73, on Mac OS X 10.15.5 19F101, locale en-GB)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 3.6)
[✓] VS Code
[✓] Connected device (2 available)

• No issues found!
```

</details>

are you experiencing the issue with the latest master?
Thank you

**Comment by erickzanardo** [comment]
@iapicca Thanks for the response.

I was on the master channel already, but to be sure, I have updated it to the latest. And the issue still happens.

By looking on your logs, it seems that you released the second key, soon after you pressed it. You will see the problem when you press the second, and keep pressing both.

Let me try to make a better explain of the logs
```
true - w - RawKeyDownEvent <- I started pressing W here, and kept pressing it
true - d - RawKeyDownEvent <- Soon after, I started pressing D here, and kept pressing it
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
false - w - RawKeyUpEvent <- This shoudn't have occored, I haven't released the W key here
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
true - d - RawKeyDownEvent
false - w - RawKeyUpEvent <- Released W
false - d - RawKeyUpEvent <- Released D
```

**Comment by iapicca** [comment]
Hi @erickzanardo 
thanks for clarifying it,
testing your code sample with the latest master 
I now experience the behaviour you are describing

<details>
  <summary>logs</summary>

```bash
🔥  To hot restart changes while running, press "r" or "R".
For a more detailed help message, press "h". To quit, press "q".
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - w - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
false - w - RawKeyUpEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
false - w - RawKeyUpEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
true - p - RawKeyDownEvent
false - p - RawKeyUpEvent

```

</details>


<details>
  <summary>doctor</summary>

```console
[✓] Flutter (Channel master, 1.19.0-4.0.pre.99, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.19.0-4.0.pre.99 at /Users/nevercode/development/flutter_master
    • Framework revision 7467492444 (41 minutes ago), 2020-06-08 02:03:01 -0400
    • Engine revision dfdd88deb7
    • Dart version 2.9.0 (build 2.9.0-13.0.dev 87c5e3612b)


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (2 available)
    • Web Server • web-server • web-javascript • Flutter Tools
    • Chrome     • chrome     • web-javascript • Google Chrome 83.0.4103.61

• No issues found!
```

</details>

thank you for reporting this

**Comment by erickzanardo** [comment]
Hi! Is there any news on this?

If I can get some basic directions on where to look for this issue, I can give this a try myself :D

Thanks

**Comment by gspencergoog** [comment]
I was able to reproduce this as well on web (chrome), but not on macOS, so my suspicion is that this is a difference in how the web processes auto-repeated key events.

@mdebbar Do you have any idea why the web might send a keyup to one of the keys when the other starts to auto-repeat?

**Comment by gspencergoog** [comment]
Also, this doesn't seem to be inherent to the way the web does events: I tried the same thing [here](https://javascript.info/keyboard-events), and it doesn't have the spurious keyup.

**Comment by erickzanardo** [comment]
This line: https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/engine/keyboard.dart#L109 seems to have something to do with the issue (thanks for pointing that @ditman on Discord)
